### PR TITLE
(Feature) Total costs calculation on LIMIT 

### LINKED ISSUE #101: Add "Total Cost" field of "Order Details" in buy/sell component (when in Limit)
Check #58 comments/updates.

### PR DESCRIPTION
**Pull request by Agupane**
Added UI implementation to display total cost on LIMIT

closes #101

### CODE REVIEW

File: src/util/tokens.ts [hunk 1]
```diff
@@ -5,6 +5,11 @@ export const tokenAmountInUnits = (amount: BigNumber, decimals: number, toFixedD
     return amount.div(decimalsPerToken).toFixed(toFixedDecimals);
 };
 
+export const tokenAmountInUnitsToBigNumber = (amount: BigNumber, decimals: number): BigNumber => {
```
Comment (fvictorio) [inline src/util/tokens.ts]: Req: `tokenAmountInUnits` could use this function and add the `toFixed` to the result.

File: src/util/market_prices.ts [hunk 2]
```diff
@@ -32,6 +38,35 @@ export const getZeroXPriceInUSD = async (): Promise<BigNumber> => {
...
+            sellOrders = sellOrders
+                .filter(order => order.side === OrderSide.Sell)
+                .sort((o1, o2) => o1.price.comparedTo(o2.price));
+            /* Check if there is any order that could fill the needs of 0x */
+            const orderWithEnoughZeroX = sellOrders.find(sellOrder => {
+                return sellOrder.size.cmp(MAKER_FEE) === 1;
+            });
+            zeroXPrice = orderWithEnoughZeroX ? orderWithEnoughZeroX.price : zeroXPrice;
+        }
+        if (!zeroXPrice) {
```
Comment (fvictorio) [inline src/util/market_prices.ts]: Q: Is there an scenario where this variable is falsy? I don't see it.

File: src/util/market_prices.ts [hunk 3]
```diff
@@ -32,6 +38,35 @@ export const getZeroXPriceInUSD = async (): Promise<BigNumber> => {
...
+        const knownTokens = getKnownTokens(networkId);
+        const zeroXToken = knownTokens.getTokenBySymbol('ZRX');
+        let sellOrders = await getAllOrdersAsUIOrders(zeroXToken);
+        if (sellOrders && sellOrders.length > 0) {
+            sellOrders = sellOrders
+                .filter(order => order.side === OrderSide.Sell)
+                .sort((o1, o2) => o1.price.comparedTo(o2.price));
+            /* Check if there is any order that could fill the needs of 0x */
+            const orderWithEnoughZeroX = sellOrders.find(sellOrder => {
+                return sellOrder.size.cmp(MAKER_FEE) === 1;
```
Comment (fvictorio) [inline src/util/market_prices.ts]: Req: This is the same than using `greaterThan`, right? If that's the case:

- Using that function should be clearer.
- Shouldn't the comparison be greater than or equal to?

File: src/util/market_prices.ts [hunk 4]
```diff
@@ -32,6 +38,35 @@ export const getZeroXPriceInUSD = async (): Promise<BigNumber> => {
...
+export const getZeroXPriceInWeth = async (): Promise<BigNumber> => {
+    /* Gets all the available sell orders of 0x and takes the lowest price, is there are no 0x sells orders available, the price is calculated using coin market */
+    const web3Wrapper = await getWeb3Wrapper();
+    let zeroXPrice = new BigNumber('0');
+    if (web3Wrapper) {
+        const networkId = await web3Wrapper.getNetworkIdAsync();
+        const knownTokens = getKnownTokens(networkId);
+        const zeroXToken = knownTokens.getTokenBySymbol('ZRX');
+        let sellOrders = await getAllOrdersAsUIOrders(zeroXToken);
+        if (sellOrders && sellOrders.length > 0) {
```
Comment (fvictorio) [inline src/util/market_prices.ts]: Q: Can `sellOrders` be falsy?

File: src/util/market_prices.ts [hunk 5]
```diff
@@ -32,6 +38,35 @@ export const getZeroXPriceInUSD = async (): Promise<BigNumber> => {
...
+            sellOrders = sellOrders
+                .filter(order => order.side === OrderSide.Sell)
+                .sort((o1, o2) => o1.price.comparedTo(o2.price));
+            /* Check if there is any order that could fill the needs of 0x */
+            const orderWithEnoughZeroX = sellOrders.find(sellOrder => {
```
Comment (fvictorio) [inline src/util/market_prices.ts]: Q: What is the reason for this? In the (unlikely) case where the cheapest sell order has a size that is less than the fee amount, why wouldn't that price be a good enough estimation?

File: src/util/market_prices.ts [hunk 6]
```diff
@@ -32,6 +38,35 @@ export const getZeroXPriceInUSD = async (): Promise<BigNumber> => {
...
+                return sellOrder.size.cmp(MAKER_FEE) === 1;
+            });
+            zeroXPrice = orderWithEnoughZeroX ? orderWithEnoughZeroX.price : zeroXPrice;
+        }
+        if (!zeroXPrice) {
+            const zeroXInUSD = await getZeroXPriceInUSD();
+            const ethInUSD = await getEthereumPriceInUSD();
+            zeroXPrice = zeroXInUSD.mul(ethInUSD);
```
Comment (fvictorio) [inline src/util/market_prices.ts]: Q: How does this work? It seems like it's multiplying a price in USD by a price in USD, does that make sense?

File: src/components/marketplace/order_details.tsx [hunk 7]
```diff
@@ -70,18 +90,120 @@ class OrderDetails extends React.Component<Props, State> {
...
+                    limitOrder: {
+                        ...this.state.limitOrder,
+                        zeroXFeeInWeth,
+                        zeroXFeeInUSD,
+                        totalCostInWeth,
+                        totalCostInUSD,
+                    },
+                });
+            } catch (error) {
+                throw error;
```
Comment (fvictorio) [inline src/components/marketplace/order_details.tsx]: Req: Catching and rethrowing an error is the same as not including the try/catch, as far as I know.

File: src/components/marketplace/order_details.tsx [hunk 8]
```diff
@@ -70,18 +90,120 @@ class OrderDetails extends React.Component<Props, State> {
...
+        },
+    };
+
+    public updateLimitOrderState = async () => {
+        const { tokenAmount, tokenPrice, selectedToken } = this.props;
+        if (selectedToken) {
+            /* Reduces decimals of the token amount */
+            const tokenAmountConverted = tokenAmountInUnitsToBigNumber(tokenAmount, selectedToken.decimals);
+            /* This could be refactored with promise all  */
+            const promisesArray = [getZeroXPriceInWeth(), getZeroXPriceInUSD(), getEthereumPriceInUSD()];
```
Comment (fvictorio) [inline src/components/marketplace/order_details.tsx]: Q: Am I missing something here or does this mean that we are going to get all the ZRX orders from the relayer every time the user presses a key?

Even if that's not a problem, shouldn't we handle the ZRX/wETH price the same way we handle the ETH/USD price: asking for it in regular intervals and having it as part of the global state?

File: src/components/marketplace/order_details.tsx [hunk 9]
```diff
@@ -70,18 +90,120 @@ class OrderDetails extends React.Component<Props, State> {
...
-        const { orderDetailType } = this.state;
+        const { orderType } = this.props;
+        let render = null;
+        if (orderType === OrderType.Limit) {
+            const { orderDetailType, zeroXFeeInUSD, totalCostInWeth, totalCostInUSD } = this.state.limitOrder;
+            render = this._renderOrderDetails(orderDetailType, zeroXFeeInUSD, totalCostInWeth, totalCostInUSD);
+        }
+        if (orderType === OrderType.Market) {
+            const { orderDetailType, zeroXFeeInUSD, totalCostInWeth, totalCostInUSD } = this.state.marketOrder;
+            render = this._renderOrderDetails(orderDetailType, zeroXFeeInUSD, totalCostInWeth, totalCostInUSD);
```
Comment (fvictorio) [inline src/components/marketplace/order_details.tsx]: Req: This line and line 183 are equal, they could be outside the `if`s.

File: src/components/marketplace/order_details.tsx (hunk 8 above)
Comment (Agupane) [reply src/components/marketplace/order_details.tsx]: It's now done as a set interval cache, we should add it on the store, maybe on the issue #77 ? @mariano-aguero

File: src/components/marketplace/order_details.tsx (hunk 9 above)
Comment (Agupane) [reply src/components/marketplace/order_details.tsx]: I'd just changed it

File: src/components/marketplace/order_details.tsx (hunk 7 above)
Comment (Agupane) [reply src/components/marketplace/order_details.tsx]: I'd removed it.

File: src/util/market_prices.ts (hunk 6 above)
Comment (Agupane) [reply src/util/market_prices.ts]: Thanks for checking this, I'd made a mistake, it's a division what is needed to calculate that, now it's changed

File: src/util/market_prices.ts (hunk 5 above)
Comment (Agupane) [reply src/util/market_prices.ts]: I'm not very sure if this is a good of wrong estimation. I'd discussed it with @unjapones and we decided to go for the "check the market price" approach, but we could just remove it and use the orderbook value. What do you guys think?

File: src/util/market_prices.ts (hunk 4 above)
Comment (Agupane) [reply src/util/market_prices.ts]: I'd removed that check

File: src/util/market_prices.ts (hunk 3 above)
Comment (Agupane) [reply src/util/market_prices.ts]: Done, now it's using that function and it's also checking the equality, thanks!

File: src/util/market_prices.ts (hunk 2 above)
Comment (Agupane) [reply src/util/market_prices.ts]: I'd changed it now :)

File: src/util/tokens.ts (hunk 1 above)
Comment (Agupane) [reply src/util/tokens.ts]: Done!

**Pull request merged**